issue-1169: Use SpringBoot "@Value + Resource" to load YAML resources.

bono007 opened the pull request:
# Pull Request Description

This is a proposed fix for #1169 

## Technical details

* Uses built-in SpringBoot `@Value + Resource` instead of custom ResourceLoader
* In order for the above to work - defined a function to be used in `SpEL` of the `@Value` to resolve path properly
* Override the path resolve in test env by setting an override property prefixed to `classpath:`

**Concerns:**
* Although we now leverage the built-in Spring Boot reosurce loading, some of the complexity around the path resolving is still present - albeit pushed under the surface a bit. I was hoping to not have to "pretty up" the paths like this but I wanted to follow through w/ the solution, get it in front of everyone as a working PR, and then we can decide what we want to do w/ it.

### TODOs
* [ ] Group to decide to move forward or not w/ these changes
* [ ] If group consensus is to move forward - create a unit test for `YamlPropertyPathResolver`.

# Acceptance Test

* [x] Building the code with `mvn clean install -Dintegration.tests` still works.
* [x] Running `mvn spring-boot:run` in the `strongbox-web-core` still starts up the application correctly.
* [ ] Building the code and running the `strongbox-distribution` from a `zip` or `tar.gz` still works.
* [ ] The tests in the [`strongbox-web-integration-tests`](https://github.com/strongbox/strongbox-web-integration-tests/) still run properly.

# Questions

* Does this pull request break backward compatibility? 
  * [ ] Yes
  * [x] No

* Does this pull request require other pull requests to be merged first? 
  * [ ] Yes, please see #...
  * [x] No

* Does this require an update of the documentation?
  * [ ] Yes, please see strongbox/strongbox-docs#{PR_NUMBER}
  * [x] No

bono007 commented:
@sbespalov
@fuss86
@carlspring

Please and thank you...

carlspring commented:
Hi, @bono007 !

Thank you for your contribution!

Would you mind signing the ICLA, as described in the [Contributing](https://github.com/strongbox/strongbox/blob/master/CONTRIBUTING.md) page?

bono007 commented:
Done. Emailed to carlspring@gmail.com

fuss86 commented:
I'll try to review it tomorrow :) Sorry

bono007 commented:
No worries. I am at SpringOne so I have been busy doing that all day.

carlspring commented:
Thanks for signing the ICLA! :)

bono007 commented:
@fuss86 
Yeh it is just an alternate impl hidden a bit below the surface. I think its worth adding as well. I like your suggestion of de-yaml'ing it and making it more generic. I will do that. I don't think I will name i t `resource loader` as that is too close to normal boot classes. I will come up w/ something though.

bono007 commented on strongbox-commons/src/main/java/org/carlspring/strongbox/config/YamlPropertyPathResolverConfig.java:
```diff
@@ -0,0 +1,81 @@
+ * @author cbono
+ */
+@Configuration
+class YamlPropertyPathResolverConfig
```
Named -> "PropertiesPath*"

fuss86 replied:
Still `YamlPropertyPathResolverConfig` which contains `Yaml` :)

bono007 commented on strongbox-commons/src/main/java/org/carlspring/strongbox/config/YamlPropertyPathResolverConfig.java:
```diff
@@ -0,0 +1,81 @@
+class YamlPropertyPathResolverConfig
+{
+    @Bean(name = "yamlPath")
+    YamlPropertyPathResolver yamlPropertyPathResolver() {
```
Done, renamed to "propsPath"

fuss86 commented on strongbox-commons/src/main/java/org/carlspring/strongbox/config/YamlPropertyPathResolverConfig.java:
```diff
@@ -0,0 +1,82 @@
+         *
+         * @param customPathPropertyName the name of the property that may hold a user specified path
+         * @param defaultPath default path to use if no user specified path is set in the custom path property
+         * @return final path to use to lookup the YAML resource - including the proper prefix ('file:' or 'classpath:')
```
> @return final path to use to lookup the YAML resource - including the proper prefix ('file:' or 'classpath:')

Still YAML related

bono007 replied:
Ah yeh - missed these files. Will update in a bit.

bono007 commented:
Uggh Windows is trash. I will look into that as well.

bono007 commented:
@fuss86 that clearly did not help. So... how do you run tests against windows?

bono007 commented:
AWS route did not workout. I found an ISO and installed into VirtualBox... I can now repro the issue... I am going to bed now but will dig more in the am. I will get it straightened out.

<img width="1360" alt="Screen Shot 2019-10-10 at 10 50 49 PM" src="https://user-images.githubusercontent.com/28907971/66623285-9189a800-ebb0-11e9-9bce-8db5cd12cce1.png">

bono007 commented:
@fuss86 ok - hopefully this one works.. I got past the issue on my local Windows VM. 🤞

bono007 commented:
@fuss86 we have a green run... I repeat... we have a green run :)

carlspring commented:
I'll spin off a build on our infractructure! Thanks! :)

steve-todorov commented:
@bono007 it looks like your changes are affecting the `strongbox-distribution`. Could you please check it out? 

You should be able to go into the `Artifacts` tab in Jenkins and download the `strongbox-distribution/strongbox-vault/logs/strongbox.log`. It contains strongbox's startup log for the distribution and a few exceptions have occurred there (there fore failing the Jenkins stage).

bono007 commented:
@carlspring can you kick off another distribution build?

carlspring commented:
@bono007 ,

The username under which you're committing and pushing to Github seems to be mis-configured. You will have to look into how to configure your git settings properly, amend the author of the commits and then force push back to the branch. (When your user is properly set up, you see your avatar when against your commits in the commit history, otherwise it's a default grey avatar).

carlspring commented:
I'm afraid it's still fIling.

bono007 commented:
Interesting on the commit history. I will look into it.

bono007 commented:
@carlspring - I squashed and rebased w/ proper user. Also, this should fix the issue. My previous fix was off by a `/`. Also, I need to add a unit test for this path resolver once this is fixed.

bono007 commented:
@carlspring 
@fuss86 

I have added a unit test to cover the property path resolver. This is ready for a final once-over and then one more round through verification.

carlspring commented:
@bono007 ,

Your latest changes are building fine across our pipeline. Thanks! :)

@fuss86, @sbespalov ,

Could you please review? Thanks! :)

carlspring commented on strongbox-commons/src/main/java/org/carlspring/strongbox/config/PropertiesPathResolverConfig.java:
```diff
@@ -0,0 +1,106 @@
+import org.springframework.util.StringUtils;
+
+/**
+ * Configures a {@link }PropertiesPathResolver} under the name "propsPath" - it can then be used in SpEL
```
Is there a missing brace here,,?

fuss86 commented on strongbox-testing/strongbox-testing-core/src/main/java/org/carlspring/strongbox/config/AutomaticTestPropertiesEnvironmentPostProcessor.java:
```diff
@@ -0,0 +1,37 @@
+    {
+        final Map<String, Object> properties = new HashMap<>();
+
+        // Force all resolved YAML paths to be prefixed with 'classpath:' during tests
```
`YAML` word not needed here

fuss86 commented on strongbox-commons/src/main/java/org/carlspring/strongbox/config/PropertiesPathResolverConfig.java:
```diff
@@ -0,0 +1,106 @@
+                return currentDirAbs + (path.length() > 1 ? path.substring(1) : "");
+            }
+
+            return path;
```
will this method ensure that `etc/conf/conf.yaml` is absolute ? If `etc/conf/conf.yaml` is provided then `etc/conf/conf.yaml` is returned which is not absolute

bono007 replied:
I have renamed it as the intention of the method is to ensure that if it begins w/ `.` or `..` (current or parent dir) that it expands those as `FileUriResource` does not handle those well. 

Need to protect against the cases where a user-specified custom properties path (via the custom property) or a `strongbox.home` has been defined w/ a `.` or `..`

bono007 commented:
@carlspring 
@fuss86 
I have addressed the latest comments - please review.

carlspring commented on strongbox-commons/src/main/java/org/carlspring/strongbox/config/PropertiesPathResolverConfig.java:
```diff
@@ -0,0 +1,106 @@
+@Configuration
+class PropertiesPathResolverConfig
+{
+    @Bean(name = "propsPath")
```
We typically don't abbreviate the names of beans like this. Would you mind using the full name (`propertiesPathResolver`)? Thanks! :)

bono007 replied:
I usually don't either. However, the whole purpose of this bean is to be used in a SpEL statement so I strove for compact name. That being said, 

```java
@Value("#{@propsPath.resolve('strongbox.cron.tasks.yaml','etc/conf/strongbox-cron-tasks.yaml')}")
```
is not much better than 
```java
@Value("#{@propertiesPathBean.resolve('strongbox.cron.tasks.yaml','etc/conf/strongbox-cron-tasks.yaml')}")
```

I will expand it,

bono007 replied:
@carlspring this is done.

carlspring replied:
Yeah, I get what you mean... For brevity, it's better but for clarity -- probably not.

Thanks! :)

bono007 commented:
When can we merge this one?

carlspring commented on strongbox-commons/src/test/java/org/carlspring/strongbox/config/PropertiesPathResolverConfigTest.java:
```diff
@@ -0,0 +1,112 @@
+    private static final String CURRENT_DIR = System.getProperty("user.dir");
+    private static final String CUSTOM_PATH_PROP = "strongbox.scary-t.file";
+    private static final String STRONGBOX_HOME_PROP = "strongbox.home";
+    private static final String STRONGBOX_HOME_DEFAULT = "/Users/keyboard-kitty/work/sbox";
```
This won't work on other people's machines... Could you please fix it? Thanks!
(Also the path should be `strongbox` and not be abbreviated).

bono007 replied:
These are mock paths, not real ones

carlspring commented on strongbox-commons/src/test/java/org/carlspring/strongbox/config/PropertiesPathResolverConfigTest.java:
```diff
@@ -0,0 +1,112 @@
+    @Test
+    public void testCustomAbsolutePath()
+    {
+        final String customPropValue = "/Users/sleepy/one/conf.yaml";
```
As above.

bono007 replied:
They are mock paths

carlspring replied:
I see! Okay, thanks for the clarification! :)

bono007 commented:
Merge? :)

carlspring commented:
> Merge? :)

We're waiting for @sbespalov 's final word. :)

sbespalov commented on strongbox-commons/src/main/java/org/carlspring/strongbox/config/PropertiesPathResolverConfig.java:
```diff
@@ -0,0 +1,106 @@
+class PropertiesPathResolverConfig
+{
+    @Bean(name = "propertiesPathBean")
+    PropertiesPathResolver propertiesPathResolver(Environment environment)
```
`propertiesPathResolver` seems better name.

carlspring replied:
I agree.

bono007 replied:
Done. Nws. I should have named it that from the very beginning anyways. Try to be clever and see where it gets ya! LOL.

bono007 commented:
Merge? :)

carlspring commented:
Thanks for all the hard work, @bono007 ! Congratulations on your first pull request! :)
